### ISSUE TITLE
Multithreading safety

### ISSUE THREAD

**Issue by gevorgter** [xamarin/TorchSharp#941]
Great library but please provide some information on multi threading safety.
In production use, models are normally wrapped into REST service which is inherently multithreaded. The Torch itself is safe to use for inference in that matter.
Not sure if your library keeping it this way. Major concern is introduction of scopes for memory management. 
Stack of scopes seems to me existing on global level which can be a problem. Not sure what else.

**Comment by NiklasGustafsson** [comment]
> For the latter, we might just need to switch to `AsyncLocal<T>`. I believe it does work with both threads and async flows.

As always an excellent point, @lostmsu. The current solution accommodates thread parallelism, but not task-based concurrency.

**Comment by NiklasGustafsson** [comment]
>> The Torch itself is safe to use for inference in that matter.

@gevorgter -- I have always wondered where, in all of the torch documentation, that safety is formalized into a guarantee. Do you have a reference?

**Comment by gevorgter** [comment]
Sorry, no reference. But my hunch is that training might not be thread safe but inference is since there is no gradient to keep and it's just calculations.

Pytorch team appears to care about threading issues and fixing them.
https://github.com/pytorch/pytorch/issues/16828

**Comment by alinpahontu2912** [comment]
@gevorgter, thank you for reporting this important issue regarding multithreading safety! We appreciate your insights on the implications of using TorchSharp in multithreaded environments, especially concerning memory management and scopes. 

Since it's been over two years, I wanted to check in to see if this issue is still relevant or if you’ve noticed any improvements in your use of the library. TorchSharp has evolved significantly during this time, and we’d love to hear if your concerns have been addressed or if there are still areas we should focus on. 

Your feedback is invaluable to us, so please let us know!
